### PR TITLE
docs: document [copr-build] keyword

### LINKED ISSUE #766: Separate configuration options for Pagure pushes and pull-requests
Original issue: https://pagure.io/copr/copr/issue/766
Opened: 2019-05-28 20:26:10
Opened by: [churchyard](https://accounts.fedoraproject.org/user/churchyard)

I've [recently realized](https://lists.fedorahosted.org/archives/list/copr-devel@lists.fedorahosted.org/thread/26ZXCMM5BNIFXLUVQ6JVEIAVUCM7VSZX/) that when I have automatic rebuilds for packages in [Fedora's scm](http://src.fedoraproject.org/), all pull requests are built as well.

I have SCM setup like this:

```
$ copr edit-package-scm --clone-url https://src.fedoraproject.org/rpms/<pkgname>.git --name <pkgname> --webhook-rebuild on --commit master <copr>
```

I want a new commit to master to trigger a build however I don't want a PR against the project to trigger a build. I consider that feature useful but not reasonable as a default behavior that cannot be changed. Also, I consider the default behavior very confusing.

---

#### [praiskup](https://accounts.fedoraproject.org/user/praiskup) commented at 2019-07-26 09:37:08:
See also https://bugzilla.redhat.com/show_bug.cgi?id=1661263

---

#### [churchyard](https://accounts.fedoraproject.org/user/churchyard) commented at 2019-07-26 10:54:03:
Yes, this seem more like a bug than a RFE. The behavior is very confusing.

---

#### [churchyard](https://accounts.fedoraproject.org/user/churchyard) commented at 2020-11-23 20:14:21:
Note that after a year, I grew fond of this feature very much. I still think an opt-out would be nice, but I would not use it.

---

#### [tstellar](https://accounts.fedoraproject.org/user/tstellar) commented at 2022-03-24 17:47:59:
I think it would be useful to have an option to disable builds triggered by pull requests, but still allow builds triggered by commits.  I have several large repos where I only want to build from official commits, and building the PRs seems to be a waste of resources.

### LINKED ISSUE #1681: Document [copr-build] keyword
Original issue: https://pagure.io/copr/copr/issue/1681
Opened: 2021-02-02 06:49:26
Opened by: [praiskup](https://accounts.fedoraproject.org/user/praiskup)

... for the pagure integration.

### PR DESCRIPTION
**Pull request by nikromen**
Fixes #1681 
Fixes #766

### CODE REVIEW

File: doc/user_documentation.rst [hunk 1]
```diff
@@ -273,12 +273,14 @@ removed after 40 days of build inactivity.
 Pagure Integration
 ------------------
 
-For any pagure instance (including `src.fedoraproject.org <http://src.fedoraproject.org/>`_), you can setup Copr auto-rebuilding and pr/commit flagging on new changes landing into a pagure repository and its open pull requests.
+For any pagure instance (including `src.fedoraproject.org <http://src.fedoraproject.org/>`_), you can setup Copr
```
Comment (praiskup) [inline doc/user_documentation.rst:276]: s/pagure/Pagure/

File: doc/user_documentation.rst [hunk 2]
```diff
@@ -293,6 +295,24 @@ of the names ``<coprname>:pr:<pr_id>``. ``<pr_id>`` is ID of the pull request op
...
+
+- Go to ``Settings`` on your project
+- Go to ``Project Options`` section
+- Select ``Fedmsg notifications`` option
+
+How to se up project in Copr:
+
+- Create new package in your Copr repository
+- Select ``SCM`` source type and specify ``Clone url``
+- Check the ``Auto-rebuild`` section
```
Comment (praiskup) [inline doc/user_documentation.rst:311]: The`[copr-build]` keyword doesn't need a specific project configuration actually.  It is automatically available with the `Auto-rebuilding` steps above.  This seems redundant, or could be at least "merged" into the steps above?

File: doc/user_documentation.rst [hunk 3]
```diff
@@ -293,6 +295,24 @@ of the names ``<coprname>:pr:<pr_id>``. ``<pr_id>`` is ID of the pull request op
...
+
+From now on, everytime you comment ``[copr-build]`` keyword into PR, Copr build is automatically triggered into
+the specified project in Copr.
```
Comment (praiskup) [inline doc/user_documentation.rst:314]: Free form (please update): From now on, you get automatic rebuilds for new pull-requests (including force pushes), you can drop a comment with `[copr-build]` keyword to re-trigger the build(s), and copr also automatically builds on pushes (commit) to the Pagure repo.

**Comment by praiskup** [comment]
Thank you for the PR!

File: doc/user_documentation.rst (hunk 2 above)
Comment (nikromen) [reply doc/user_documentation.rst:311]: I did some research and the `auto-rebuilding` part and `PR-comment` rebuilding part really differ.  If you disable the `Fedmsg` in `Hooks` then `Auto-rebuilding` stop working. And if you disable ` Fedmsg notifications ` in `Project option` then PR-comment rebuild stops working. These really are two independent settings.

Also from `Fedmsg` option in `Hooks`:
```
This hook pushes commit notification to the fedmsg bus to be consumed by other
applications.

It is different from the fedmsg setting present in the project options block
which publishes notifications about events happening in the project via
pagure's (web) user interface, for example: new tickets, new comments, new
pull-requests and so on.
This hook on the other hand only acts on commits.
```

So what is says is the one option is really for Pagure-UI related things (things you do in browser) and the other is for commits etc.

File: doc/user_documentation.rst (hunk 3 above)
Comment (nikromen) [reply doc/user_documentation.rst:314]: this describes impact from auto-rebuild and PR command rebuild together but as I described in https://github.com/fedora-copr/copr/pull/2662#discussion_r1167479095 these two really differs in the settings to enable. Seems like there were some changes in Pagure recently

Comment (praiskup) [reply doc/user_documentation.rst:314]: I still don't get how these two are actually different :-) I likely miss something and I need yet another video-meeting iteration.  The point of this PR should be to have a self-contained single howto (not two howtos).

**Comment by nikromen** [comment]
@praiskup ptal

File: doc/user_documentation.rst [hunk 4]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
...
+remote Pagure repository, see :ref:`scm_ref` if you need more detailed settings. Also make sure,
+"Auto-rebuild" checkbox is checked.
 
-Now your SCM package will get rebuilt on new commits into the main repo as well as into open PRs.
+Now your SCM package will get rebuilt on new commits into the main repo.
 
-Note that built changes coming from pull requests are not actually placed into the main copr repository. Instead, they are being placed into side repositories
-of the names ``<coprname>:pr:<pr_id>``. ``<pr_id>`` is ID of the pull request opened in Pagure. On Fedora, you can enable the side repository to test the changes with:
+Note that built changes coming from pull requests are not actually placed into the main copr repository.
+Instead, they are being placed into side repositories of the names ``<coprname>:pr:<pr_id>``. ``<pr_id>``
```
Comment (praiskup) [inline doc/user_documentation.rst:294]: some remaining typo here

File: doc/user_documentation.rst [hunk 5]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
 Pagure Integration
 ------------------
 
-For any pagure instance (including `src.fedoraproject.org <http://src.fedoraproject.org/>`_), you can setup Copr auto-rebuilding and pr/commit flagging on new changes landing into a pagure repository and its open pull requests.
+For any Pagure instance (including `src.fedoraproject.org <http://src.fedoraproject.org/>`_), you can setup Copr
+auto-rebuilding, triggering build on a PR comment containing ``[copr-build]`` keyword and pr/commit flagging on new
+changes landing into a pagure repository and its open pull requests.
```
Comment (praiskup) [inline doc/user_documentation.rst:278]: This intro is very hard to read and understand, after the years I'd propose simplifying it:

You can configure Copr to do automatic builds for Pagure projects. Builds can be triggered by commits pushed and/or newly opened pull requests.

File: doc/user_documentation.rst [hunk 6]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
...
 
 Auto-rebuilding
 ^^^^^^^^^^^^^^^
 
-On the Pagure side, you need to set Fedmsg to 'active' in your project settings (in 'Hooks' section almost at the bottom). For some instances
-(e.g. `src.fedoraproject.org <http://src.fedoraproject.org/>`_), this might already be active by default so you don't even need to perform this step.
+To enable auto-rebuilding on push events, you need to set Fedmsg to 'active' in your project settings on Pagure
+(in 'Hooks' section). For some instances (e.g. `src.fedoraproject.org <http://src.fedoraproject.org/>`_), this
```
Comment (praiskup) [inline doc/user_documentation.rst:284]: s/in Hooks/in the Hooks/

File: doc/user_documentation.rst [hunk 7]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
...
+might already be active by default so you don't even need to perform this step.
```
Comment (praiskup) [inline doc/user_documentation.rst:285]: s/already be active by default so you don't even need to perform this step/already be active for all projects./

File: doc/user_documentation.rst [hunk 8]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
...
 
-In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the remote Pagure repository, see :ref:`scm_ref`
-if you need more detailed settings. Also make sure, "Auto-rebuild" checkbox is checked.
+In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the
+remote Pagure repository, see :ref:`scm_ref` if you need more detailed settings. Also make sure,
+"Auto-rebuild" checkbox is checked.
 
-Now your SCM package will get rebuilt on new commits into the main repo as well as into open PRs.
+Now your SCM package will get rebuilt on new commits into the main repo.
```
Comment (praiskup) [inline doc/user_documentation.rst:291]: I'd personally drop this statement?

File: doc/user_documentation.rst [hunk 9]
```diff
@@ -273,26 +273,37 @@ removed after 40 days of build inactivity.
...
 
-Note that built changes coming from pull requests are not actually placed into the main copr repository. Instead, they are being placed into side repositories
-of the names ``<coprname>:pr:<pr_id>``. ``<pr_id>`` is ID of the pull request opened in Pagure. On Fedora, you can enable the side repository to test the changes with:
+Note that built changes coming from pull requests are not actually placed into the main copr repository.
```
Comment (praiskup) [inline doc/user_documentation.rst:293]: s/Note that built changes... /Note that pull request build results are placed into side directories containing pull-request IDs, like `<project_name>:pr:<pull_request_id>`.

**Comment by nikromen** [comment]
Yes you are right, this covers the #766

File: doc/user_documentation.rst [hunk 10]
```diff
@@ -273,21 +273,28 @@ removed after 40 days of build inactivity.
...
+To enable auto-rebuilding on push events, you need to set Fedmsg to 'active' in your project settings on Pagure
+(in the 'Hooks' section). For some instances (e.g. `src.fedoraproject.org <http://src.fedoraproject.org/>`_), this
+might already be active for all projects.
 
-In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the remote Pagure repository, see :ref:`scm_ref`
-if you need more detailed settings. Also make sure, "Auto-rebuild" checkbox is checked.
+To enable triggering builds on PRs, on Pagure go to ``Settings`` -> ``Project Options`` and select the
+``Fedmsg notifications`` option. From now on, you get automatic rebuilds for new pull-requests (including
+force pushes), you can drop a comment with ``[copr-build]`` keyword to re-trigger the builds, and Copr also
+automatically builds on pushes (commit) to the Pagure repo.
```
[the remark this answers is not in the document]
Comment (praiskup) [reply doc/user_documentation.rst:289]: Well ... or drop the part `... builds on pushes` suffix?

**Comment by praiskup** [comment]
Would you mind documenting the "badge flagging" feature?  Or do we have it somewhere?
- generate an API key in Pagure
- use it in Integration tab in the corresponding Copr project?

**Comment by praiskup** [comment]
#2675 is an attempt to complement this PR. Lemme know which one should go in first; depending on that we should also consider moving the Pagure Integration part into a separate page (because the main `user_documentation` one is too big.

File: doc/user_documentation.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (nikromen) [reply doc/user_documentation.rst:289]: It should be a separate paragraph but the newline got lost somewhere :)

File: doc/user_documentation.rst [hunk 11]
```diff
@@ -273,21 +273,30 @@ removed after 40 days of build inactivity.
...
 
-In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the remote Pagure repository, see :ref:`scm_ref`
-if you need more detailed settings. Also make sure, "Auto-rebuild" checkbox is checked.
+To enable triggering builds on PRs, on Pagure go to ``Settings`` -> ``Project Options`` and select the
+``Fedmsg notifications`` option.
 
-Now your SCM package will get rebuilt on new commits into the main repo as well as into open PRs.
+From now on, you get automatic rebuilds for new pull-requests (including
+force pushes), you can drop a comment with ``[copr-build]`` keyword to re-trigger the builds, and Copr also
+automatically builds on pushes (commit) to the Pagure repo.
```
Comment (praiskup) [inline doc/user_documentation.rst:291]: Hm, actually, I think we should swap this and the following paragraph.  "From now on..." isn't correct, you still have to define the Package in Copr.

File: doc/user_documentation.rst [hunk 12]
```diff
@@ -273,21 +273,28 @@ removed after 40 days of build inactivity.
...
 
-In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the remote Pagure repository, see :ref:`scm_ref`
-if you need more detailed settings. Also make sure, "Auto-rebuild" checkbox is checked.
+To enable triggering builds on PRs, on Pagure go to ``Settings`` -> ``Project Options`` and select the
+``Fedmsg notifications`` option. From now on, you get automatic rebuilds for new pull-requests (including
+force pushes), you can drop a comment with ``[copr-build]`` keyword to re-trigger the builds, and Copr also
+automatically builds on pushes (commit) to the Pagure repo.
 
-Now your SCM package will get rebuilt on new commits into the main repo as well as into open PRs.
+In Copr, you need an SCM package definition, which may be as simple as specifying a public clone URL of the
```
[the remark this answers is not in the document]
Comment (praiskup) [reply doc/user_documentation.rst:293]: I think @hroncok meant the separate "DistGit" (fedora in particular) method (not only "SCM").

**Review by praiskup — APPROVED** [review]
Thank you!

**Pull request merged**